File: submission/src/workflows/discovery_translational_science.py
"""Workflow E — discovery/translational science support. ADDITIONAL,
OPTIONAL SCOPE, not one of the three mandated workflows
(`case/INTEGRATED_CASE.md` §4). See
`submission/artefacts/WORKFLOW_E_DISCOVERY_TRANSLATIONAL_SCIENCE.md` for
the full scope statement — this does not change RUB-08 scoring.

Decision-support only: never accepts/rejects an assay result as valid
despite an instrument-firmware/reagent-qualification conflict, never
approves a translational model for portfolio use, never certifies
preclinical image data as authentic/non-manipulated, never promotes a
research-unqualified model to decision-grade status, never resolves a
target-validation disagreement between internal and licensed-dataset
evidence.

Specs closed: `submission/tests/test_prohibited_discovery_translational_science.py`.
Contract: `submission/evaluation/contracts/discovery_response.schema.json`.
Design basis: `submission/artefacts/04-ddd/domain_model.md`-analogous
invariants INV-14..18 (defined in the Workflow E artefact, not the
immutable domain_model.md).
"""
from submission.src.services.authorization import check_authorization


def assemble_discovery_response(request):
    request_id = request.get("request_id", "")
    subject_ref = request.get("subject_ref", "")
    as_of = request.get("as_of", "")
    auth_input = request.get("authorization", {})

    authorization = check_authorization({
        "request_id": request_id,
        "as_of": as_of,
        "purpose": auth_input.get("purpose", ""),
        "entitlement": request.get("entitlement", {}),
    })

    contradictions = []
    gaps = []
    required_reviews = []

    # INV-14 (INJ-007): an assay result run on an instrument whose firmware
    # does not match its qualified firmware, or with a reagent lot whose
    # CoA is not verified (e.g. transcribed_only) or expired, is never
    # accepted/rejected as valid — it is surfaced as a contradiction.
    assay_quality_flags = []
    for result in request.get("assay_results", []):
        instrument = result.get("instrument_info") or {}
        reagent = result.get("reagent_lot_info") or {}
        reasons = []
        firmware = instrument.get("firmware")
        qualified_firmware = instrument.get("qualified_firmware")
        if firmware is not None and qualified_firmware is not None and firmware != qualified_firmware:
            reasons.append("instrument_firmware_not_qualified_firmware")
        if instrument.get("qualification_status") not in (None, "qualified"):
            reasons.append(f"instrument_qualification_status={instrument.get('qualification_status')}")
        coa_status = reagent.get("coa_status")
        if coa_status not in (None, "verified"):
            reasons.append(f"reagent_coa_status={coa_status}")
        expiry = reagent.get("expiry")
        as_of_date = request.get("as_of_date")
        if expiry is not None and as_of_date is not None and expiry < as_of_date:
            reasons.append("reagent_lot_expired")
        if reasons:
            assay_quality_flags.append({
                "assay_id": result.get("assay_id"),
                "compound_code": result.get("compound_code"),
                "instrument": instrument.get("instrument_id"),
                "reagent_lot": reagent.get("reagent_lot"),
                "reasons": reasons,
            })
            contradictions.append({
                "type": "assay_result_qualification_conflict",
                "assay_id": result.get("assay_id"),
                "reasons": reasons,
            })
            required_reviews.append(f"assay_qualification_review:{result.get('assay_id')}")

    # INV-15 (INJ-009): a model performance slice showing a large subgroup
    # disparity (e.g. AUROC gap across ancestry-group slices, or a small-n
    # slice) is never silently averaged away — surfaced as a contradiction,
    # never resolved to a single portfolio-ranking figure.
    for perf in request.get("model_performance_slices", []):
        model_id = perf.get("model_id")
        slices = perf.get("slices", [])
        values = {s.get("slice"): s.get("value") for s in slices if s.get("value") is not None}
        if len(values) > 1:
            spread = max(values.values()) - min(values.values())
            if spread >= 0.10:
                contradictions.append({
                    "type": "model_performance_subgroup_disparity",
                    "model_id": model_id,
                    "metric": slices[0].get("metric") if slices else None,
                    "values_by_slice": values,
                    "spread": round(spread, 4),
                })
                required_reviews.append(f"subgroup_fairness_review:{model_id}")
        cohorts = perf.get("cohorts", [])
        for cohort in cohorts:
            if cohort.get("n") is not None and int(cohort["n"]) < 100:
                gaps.append({
                    "gap_type": "small_cohort_evidence",
                    "model_id": model_id,
                    "cohort": cohort.get("cohort"),
                    "ancestry_group": cohort.get("ancestry_group"),
                    "n": cohort.get("n"),
                })
                required_reviews.append(f"cohort_size_review:{model_id}:{cohort.get('cohort')}")

    # INV-16 (INJ-010): preclinical image-forensics evidence indicating a
    # high similarity score between panels claimed as distinct is never
    # certified as authentic or dismissed as manipulated — surfaced as a
    # contradiction requiring human forensic review.
    for forensic in request.get("image_forensics", []):
        score = forensic.get("similarity_score")
        if score is not None and float(score) >= 0.90:
            contradictions.append({
                "type": "preclinical_image_similarity_concern",
                "study_id": forensic.get("study_id"),
                "panel": forensic.get("panel"),
                "similarity_to": forensic.get("similarity_to"),
                "similarity_score": score,
                "metadata_note": forensic.get("metadata_note"),
            })
            required_reviews.append(f"image_forensics_review:{forensic.get('study_id')}:{forensic.get('panel')}")

    # INV-17 (INJ-011): a model registry entry whose status is not
    # decision-grade (e.g. research_unqualified) but is referenced by
    # performance/portfolio evidence is never promoted — surfaced as a
    # gap, never a "qualified"/"approved" status.
    for entry in request.get("model_registry_entries", []):
        status = entry.get("status")
        if status not in (None, "qualified", "approved", "production"):
            gaps.append({
                "gap_type": "model_not_decision_grade",
                "model_id": entry.get("model_id"),
                "status": status,
                "intended_use": entry.get("intended_use"),
            })
            required_reviews.append(f"model_qualification_review:{entry.get('model_id')}")

    # INV-18 (INJ-012): disagreeing target-validation evidence sources
    # (internal vs. licensed-dataset) are never resolved to a single
    # direction — surfaced as a contradiction. A license whose
    # model_training permission is restricted/unclear is surfaced as a
    # gap, never silently used.
    for target in request.get("target_evidence", []):
        by_target = {}
        for row in target.get("rows", []):
            by_target.setdefault(row.get("target"), set()).add(row.get("direction"))
        for tgt, directions in by_target.items():
            if len(directions) > 1:
                contradictions.append({
                    "type": "target_evidence_disagreement",
                    "target": tgt,
                    "directions": sorted(directions),
                })
                required_reviews.append(f"target_validation_review:{tgt}")

    for license_row in request.get("data_licenses", []):
        model_training = license_row.get("model_training")
        if model_training in ("restricted", "unclear"):
            gaps.append({
                "gap_type": "data_license_training_use_unresolved",
                "dataset": license_row.get("dataset"),
                "permitted_use": license_row.get("permitted_use"),
                "model_training": model_training,
            })
            required_reviews.append(f"data_license_review:{license_row.get('dataset')}")

    if (not request.get("assay_results") and not request.get("model_performance_slices")
            and not request.get("image_forensics") and not request.get("model_registry_entries")
            and not request.get("target_evidence") and not request.get("data_licenses")
            and not request.get("evidence")):
        gaps.append({"gap_type": "no_evidence_provided", "subject_ref": subject_ref})

    return {
        "request_id": request_id,
        "workflow": "discovery_translational_science",
        "as_of": as_of,
        "authorization": authorization,
        "evidence": list(request.get("evidence", [])),
        "contradictions": contradictions,
        "gaps": gaps,
        "abstentions": list(request.get("abstentions", [])),
        "human_review": {"required": True, "role": "Translational Science Lead / Discovery QA"},
        "execution_status": "not_executed",
        "audit": {"event_id": f"AUD-{request_id}"},
        "subject_ref": subject_ref,
        "assay_quality_flags": assay_quality_flags,
        "required_reviews": required_reviews,
    }

File: submission/app-advanced/lib/workflows/discovery_translational_science.ts
// Workflow E: Discovery/Translational Science (INV-14..18). Ported from
// submission/app/app.js assembleDiscoveryResponse / DISCOVERY_SCENARIOS.
// ADDITIONAL, OPTIONAL SCOPE. Never renders assay_disposition, model_approval,
// image_authenticity, model_status_change or target_validation_conclusion.
import { Authorization, EvidenceRef, checkAuthorization } from "./common";

export interface InstrumentInfo {
  instrument_id: string;
  firmware: string;
  qualified_firmware: string;
  qualification_status: string;
}

export interface ReagentLotInfo {
  reagent_lot: string;
  coa_status: string;
  expiry: string;
}

export interface AssayResult {
  assay_id: string;
  compound_code: string;
  instrument_info: InstrumentInfo;
  reagent_lot_info: ReagentLotInfo;
}

export interface ModelPerformanceSlice {
  model_id: string;
  slices: { slice: string; metric: string; value: number }[];
}

export interface ModelRegistryEntry {
  model_id: string;
  intended_use: string;
  status: string;
}

export interface DiscoveryScenario {
  subject_ref: string;
  evidence: EvidenceRef[];
  assay_results: AssayResult[];
  model_performance_slices: ModelPerformanceSlice[];
  model_registry_entries: ModelRegistryEntry[];
}

export const DISCOVERY_SCENARIOS: Record<string, DiscoveryScenario> = {
  assayAndSubgroup: {
    subject_ref: "BX-17",
    evidence: [
      { source: "data/assay_results.csv", record_id: "AS-101" },
      { source: "data/model_performance.csv", record_id: "TRN-OMICS-2" },
    ],
    assay_results: [
      {
        assay_id: "AS-101",
        compound_code: "BX-17",
        instrument_info: { instrument_id: "INS-03", firmware: "4.8.1", qualified_firmware: "4.7.9", qualification_status: "conditional" },
        reagent_lot_info: { reagent_lot: "RG-78", coa_status: "transcribed_only", expiry: "2026-10-31" },
      },
    ],
    model_performance_slices: [
      { model_id: "TRN-OMICS-2", slices: [{ slice: "Group-A", metric: "AUROC", value: 0.86 }, { slice: "Group-B", metric: "AUROC", value: 0.61 }] },
    ],
    model_registry_entries: [],
  },
  unqualifiedModel: {
    subject_ref: "TRN-OMICS-2",
    evidence: [{ source: "data/model_registry.csv", record_id: "TRN-OMICS-2" }],
    assay_results: [],
    model_performance_slices: [],
    model_registry_entries: [{ model_id: "TRN-OMICS-2", intended_use: "portfolio ranking", status: "research_unqualified" }],
  },
  clean: { subject_ref: "", evidence: [], assay_results: [], model_performance_slices: [], model_registry_entries: [] },
};

export interface DiscoveryResponse {
  workflow: "discovery_translational_science";
  authorization: Authorization;
  subject_ref: string;
  evidence: EvidenceRef[];
  contradictions: Record<string, unknown>[];
  gaps: { gap_type: string; model_id?: string; status?: string; intended_use?: string; subject_ref?: string }[];
  abstentions: unknown[];
  assay_quality_flags: { assay_id: string; compound_code: string; instrument: string; reagent_lot: string; reasons: string[] }[];
  required_reviews: string[];
  human_review: { required: true; role: string };
  execution_status: "not_executed";
}

export function assembleDiscoveryResponse(scenarioKey: string): DiscoveryResponse {
  const s = DISCOVERY_SCENARIOS[scenarioKey];
  const authorization = checkAuthorization("discovery_translational_science");
  const contradictions: Record<string, unknown>[] = [];
  const gaps: DiscoveryResponse["gaps"] = [];
  const assay_quality_flags: DiscoveryResponse["assay_quality_flags"] = [];
  const required_reviews: string[] = [];

  // INV-14: firmware/CoA/expiry qualification conflicts surfaced, never accepted/rejected.
  for (const result of s.assay_results) {
    const instrument = result.instrument_info || ({} as InstrumentInfo);
    const reagent = result.reagent_lot_info || ({} as ReagentLotInfo);
    const reasons: string[] = [];
    if (instrument.firmware != null && instrument.qualified_firmware != null && instrument.firmware !== instrument.qualified_firmware) {
      reasons.push("instrument_firmware_not_qualified_firmware");
    }
    if (instrument.qualification_status != null && instrument.qualification_status !== "qualified") {
      reasons.push(`instrument_qualification_status=${instrument.qualification_status}`);
    }
    if (reagent.coa_status != null && reagent.coa_status !== "verified") {
      reasons.push(`reagent_coa_status=${reagent.coa_status}`);
    }
    if (reasons.length) {
      assay_quality_flags.push({
        assay_id: result.assay_id,
        compound_code: result.compound_code,
        instrument: instrument.instrument_id,
        reagent_lot: reagent.reagent_lot,
        reasons,
      });
      contradictions.push({ type: "assay_result_qualification_conflict", assay_id: result.assay_id, reasons });
      required_reviews.push(`assay_qualification_review:${result.assay_id}`);
    }
  }

  // INV-15: subgroup performance disparity surfaced, never silently averaged away.
  for (const perf of s.model_performance_slices) {
    const values: Record<string, number> = {};
    for (const sl of perf.slices) if (sl.value != null) values[sl.slice] = sl.value;
    const vals = Object.values(values);
    if (vals.length > 1) {
      const spread = Math.max(...vals) - Math.min(...vals);
      if (spread >= 0.1) {
        contradictions.push({
          type: "model_performance_subgroup_disparity",
          model_id: perf.model_id,
          metric: perf.slices[0] && perf.slices[0].metric,
          values_by_slice: values,
          spread: Math.round(spread * 10000) / 10000,
        });
        required_reviews.push(`subgroup_fairness_review:${perf.model_id}`);
      }
    }
  }

  // INV-17: research-unqualified registry entries surfaced as a gap, never promoted.
  for (const entry of s.model_registry_entries) {
    if (entry.status != null && !["qualified", "approved", "production"].includes(entry.status)) {
      gaps.push({ gap_type: "model_not_decision_grade", model_id: entry.model_id, status: entry.status, intended_use: entry.intended_use });
      required_reviews.push(`model_qualification_review:${entry.model_id}`);
    }
  }

  if (!s.assay_results.length && !s.model_performance_slices.length && !s.model_registry_entries.length && !s.evidence.length) {
    gaps.push({ gap_type: "no_evidence_provided", subject_ref: s.subject_ref });
  }

  return {
    workflow: "discovery_translational_science",
    authorization,
    subject_ref: s.subject_ref,
    evidence: s.evidence,
    contradictions,
    gaps,
    abstentions: [],
    assay_quality_flags,
    required_reviews,
    human_review: { required: true, role: "Translational Science Lead / Discovery QA" },
    execution_status: "not_executed",
  };
}
